runtime/storage.go

```diff
@@ -61,201 +55,101 @@ func NewStorage(
-	}
-
-	// Existing data, decode and keep in cache
-
-	var readStorable atree.Storable
-
-	decoder := interpreter.CBORDecMode.NewByteStreamDecoder(storedData)
+func (s *Storage) loadExistingStorageMap(address atree.Address, storageIndex atree.StorageIndex) *interpreter.StorageMap {
 
-	s.reportMetric(
-		func() {
-			readStorable, err = interpreter.DecodeStorable(decoder, atree.StorageIDUndefined)
-		},
-		func(metrics Metrics, duration time.Duration) {
-			metrics.ValueDecoded(duration)
-		},
-	)
-	if err != nil {
-		panic(err)
-	}
-
-	s.readCache[storageKey] = readStorable
-
-	return readStorable
-}
-
-func (s *Storage) WriteValue(
-	inter *interpreter.Interpreter,
-	address common.Address,
-	key string,
-	value interpreter.OptionalValue,
-) {
-	storageKey := interpreter.StorageKey{
+	storageID := atree.StorageID{
 		Address: address,
-		Key:     key,
+		Index:   storageIndex,
 	}
 
-	// Remove existing value, if any
-
-	existingStorable := s.readStorable(storageKey)
-	if existingStorable != nil {
-		interpreter.StoredValue(existingStorable, s).
-			DeepRemove(inter)
-		inter.RemoveReferencedSlab(existingStorable)
-	}
+	return interpreter.NewStorageMapWithRootID(s, storageID)
+}
 
-	// Get storable representation for new value
+func (s *Storage) storeNewStorageMap(address atree.Address, domain string) *interpreter.StorageMap {
+	storageMap := interpreter.NewStorageMap(s, address)
 
-	var newStorable atree.Storable
+	storageIndex := storageMap.StorageID().Index
 
-	switch typedValue := value.(type) {
-	case *interpreter.SomeValue:
-		var err error
-		newStorable, err = typedValue.Value.Storable(
-			s,
-			atree.Address(address),
-			// NOTE: we already allocate a register for the account storage value,
-			// so we might as well store all data of the value in it, if possible,
-			// e.g. for a large immutable value.
-			//
-			// Using a smaller number would only result in an additional register
-			// (account storage register would have storage ID storable,
-			// and extra slab / register would contain the actual data of the value).
-			math.MaxUint64,
-		)
-		if err != nil {
-			panic(err)
-		}
-
-	case interpreter.NilValue:
-		break
-
-	default:
-		panic(errors.NewUnreachableError())
+	storageKey := interpreter.StorageKey{
+		Address: common.Address(address),
+		Key:     domain,
 	}
 
-	// Only write locally.
-	// The value is eventually written back through the runtime interface in `Commit`.
+	s.writes[storageKey] = storageIndex
 
-	s.writes[storageKey] = newStorable
+	return storageMap
 }
 
 func (s *Storage) recordContractUpdate(
 	inter *interpreter.Interpreter,
 	address common.Address,
-	key string,
+	name string,
 	contract interpreter.Value,
 ) {
-	storageKey := interpreter.StorageKey{
+	key := interpreter.StorageKey{
 		Address: address,
-		Key:     key,
+		Key:     name,
 	}
 
 	// Remove existing, if any
 
-	existingStorable, ok := s.contractUpdates[storageKey]
+	existingStorable, ok := s.contractUpdates[key]
 	if ok {
```

Comment:
Yes, great catch! 

`contractUpdates` explicitly keeps `nil` entries to record contract removals, so in the case of a transaction removing a contract and adding a contract back in the same transaction, this will fail.

I've fixed this and added a test case in 7cce3ae769b5cf2b4eb1b62ec9f8d5506400be31